### PR TITLE
perf: initial version of 1M operations test

### PR DESCRIPTION
**Pull request by sergos**
The test can be used for regression testing. It is advisable to tune the machine: check the NUMA configuration, fix the pstate or similar CPU autotune. Although, running dozen times gives more-less stable result for the peak performance, that should be enough for regression identification.

### CODE REVIEW

File: perf/lua/1mops_write.lua
Comment (igormunkin) [inline perf/lua/1mops_write.lua:30]: It would be nice to tweak the benchmark setup via environment variables instead of the CLI options to run it within container.

File: perf/lua/1mops_write.lua
Comment (igormunkin) [inline perf/lua/1mops_write.lua:91]: I believe you can simply use [os.tmpname](https://www.lua.org/manual/5.1/manual.html#pdf-os.tmpname) here, instead of such kind of cleanup?

File: perf/lua/1mops_write.lua
Comment (igormunkin) [inline perf/lua/1mops_write.lua:257]: It would be perfect to have an option for TSV/CSV output to process the result via Perl script or load it into Excel to calculate some statistics.

**Review by igormunkin — CHANGES_REQUESTED** [review]
@sergos, thanks for finally moving this to the trunk! I added myself to the reviewers list, since I use this benchmark a lot (though it's [mine fork](https://gist.github.com/igormunkin/e956396d8fb3348d3ebbad60588af297)).

See my major comments for the first iteration of review below.

File: perf/lua/1mops_write.lua
Comment (sergos) [reply perf/lua/1mops_write.lua:30]: Hmm... I suppose running it in any form of virtual environment will hurt the stability of the test.
Nevertheless - I can handle it, perhaps along with rewrite the arguments parsing through the internal.argparse.

File: perf/lua/1mops_write.lua
Comment (sergos) [reply perf/lua/1mops_write.lua:91]: Then it will pile up the temporary files from sequential runs? At least for those crashed, where I can't handle all type of exits or make any on_shutdown() operable. Removing the work_dir right after box_cfg() will work only until xlog rotation, where new file should be open in this folder.

Comment (igormunkin) [reply perf/lua/1mops_write.lua:91]: What are the critical points where the benchmark can crash?

Comment (sergos) [reply perf/lua/1mops_write.lua:91]: Well, beyond other issues, the very beginning of the test
```{lua}
[s.ostanevich@dev1 lua]$ ../../src/tarantool -i
tarantool> box.cfg{work_dir='asdf'}
can't chdir to `asdf': No such file or directory
[s.ostanevich@dev1 lua]$ echo $?
1
[...]
2023-07-26 13:27:47.084 [52922] main/103/interactive F> failed to preallocate 4398046511104 bytes: Cannot allocate memory, check option 'memtx_memory' in box.cfg(..)
[s.ostanevich@dev1 lua]$ echo $?
1
```
Literally - any problem in box.cfg{} can lead to an immediate termination.

Comment (sergos) [reply perf/lua/1mops_write.lua:91]: Unfortunately, the `os.tmpname()` creates a temporary file, so as `io.tmpfile()`. Apparently, there's no safe interface to provide a unique name that will be available and unique over time between its creation and using it.
The only w/a I can envision is to create the temp file, then extend its name with a suffix and create a temp dir along with it, cleaning up both upon completion. Upvote if it is feasible.

Comment (sergos) [reply perf/lua/1mops_write.lua:91]: Nevermind, will follow @ligurio proposal of `fio.tmpdir()`

File: perf/lua/1mops_write.lua
Comment (sergos) [reply perf/lua/1mops_write.lua]: Followed the rules of [compare.lua](https://github.com/tarantool/tarantool/blob/master/perf/lua/compare.lua) - see the comment below

File: perf/lua/1mops_write.lua
Comment (sergos) [reply perf/lua/1mops_write.lua]: done

File: perf/lua/1mops_write.lua
Comment (sergos) [reply perf/lua/1mops_write.lua]: done

File: perf/lua/1mops_write.lua
Comment (sergos) [reply perf/lua/1mops_write.lua]: done

File: perf/lua/1mops_write.lua
Comment (sergos) [reply perf/lua/1mops_write.lua:205]: an intentional emphasis

File: perf/lua/1mops_write.lua
Comment (sergos) [reply perf/lua/1mops_write.lua]: done, we also agreed with @locker on the `#` symbol as a comment in the tests output to let `compare.lua` to ignore them

File: perf/lua/1mops_write.lua
Comment (sergos) [reply perf/lua/1mops_write.lua]: done, using the comment prefix (see above)

File: perf/lua/1mops_write.lua
Comment (ligurio) [reply perf/lua/1mops_write.lua]: Nice! Thanks!

File: perf/lua/1mops_write.lua
Comment (ligurio) [reply perf/lua/1mops_write.lua]: Nice!

File: perf/lua/1mops_write.lua
Comment (ligurio) [inline perf/lua/1mops_write.lua:1]: Currently it is not possible to run a script without specifying interpreter:
```
$ ./perf/lua/1mops_write.lua
-bash: ./perf/lua/1mops_write.lua: Permission denied    
```

I would add a shebang:

```diff
--- a/perf/lua/1mops_write.lua
+++ b/perf/lua/1mops_write.lua
@@ -1,3 +1,5 @@
+#!/bin/env tarantool
+
 local err, popen = pcall(require,'popen')
 if err ~= true then
     print(err)
```

and give exec permissions:

```
$ chmod +x perf/lua/1mops_write.lua 
$ ./perf/lua/1mops_write.lua 
# making 10000000 REPLACE operations,
...
```

**Review by ligurio — APPROVED** [review]
LGTM (after answering to a comment about shebang)

File: perf/lua/1mops_write.lua
Comment (igormunkin) [inline perf/lua/1mops_write.lua:52]: Nit: I'd rather use `exit` helper here either, however, this is totally minor, so feel free to ignore.

File: perf/lua/1mops_write.lua
Comment (igormunkin) [inline perf/lua/1mops_write.lua:223]: Nit: It also would be nice to have an option whether run in a "dirty"/"clean" environment or even with GC disabled.

**Review by igormunkin — APPROVED** [review]
@sergos, thanks for the changes! Honestly, I see no reason to block this PR from my side. However, I left a couple of nits regarding the benchmark usage: you can fix it now, or file a ticket to improve the bench later, or even resolve it as irrelevant (if you consider anything you find below).

All in all, the changes LGETM, thanks again for moving them into the trunk!

**Comment by ligurio** [comment]
Probably it is worth fixing name and email:

> commit cb4580cf09d225e6956d8d60f7092b7c885e53a1 (sergos/sergos/1mops_perf_test, sergos/1mops_perf_test)
> Author: s.ostanevich <s.ostanevich@dev1.tarantool.i>
> Date:   Tue Jul 25 18:45:01 2023 +0300
>
>    perf: initial version of 1M operations test

File: perf/lua/1mops_write.lua
Comment (sergos) [reply perf/lua/1mops_write.lua]: Done.

File: perf/lua/1mops_write.lua
Comment (sergos) [reply perf/lua/1mops_write.lua:140]: Actually I have a limit to the number of raws in the table, it is a 1M. Should I minimize/automate the amount of memory is questionable, better make yet another CLI option with a default?

File: perf/lua/1mops_write.lua
Comment (sergos) [reply perf/lua/1mops_write.lua]: Done.

File: perf/lua/1mops_write.lua
Comment (sergos) [reply perf/lua/1mops_write.lua]: The reason is delivery(?) team claimed that Lua should have `'` to allow transparent `"` from external calls, like `-e "local s = 'hello' print(s)"`
Anyways, here it goes.

File: perf/lua/1mops_write.lua
Comment (sergos) [reply perf/lua/1mops_write.lua:160]: The replica folders are inside the temporary folder, no need to generate more of them.

File: perf/lua/1mops_write.lua
Comment (sergos) [reply perf/lua/1mops_write.lua]: Done.

File: perf/lua/1mops_write.lua
Comment (sergos) [reply perf/lua/1mops_write.lua:233]: I believe it should be eliminated as a class of possible fluctuation - these GC triggers were introduced to stabilize the results, still they failed. I would like to keep them as is, proceeding with further search of the fluctuation source.
Overloading the test with big number of options is no good IMHO.

**Pull request merged**
